### PR TITLE
Document Jenkins upgrade and changes

### PR DESCRIPTION
**Pull request by iranzo**
Document Jenkins Upgrade and changes

Preview at: https://deploy-preview-402--kubevirtio.netlify.com/2019/jenkins-ci-server-upgrade-and-jobs-for-kubevirt.html

### CODE REVIEW

File: _posts/2019-11-22-jenkins-ci-server-upgrade-and-jobs-for-kubevirt.markdown [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+In the article [Jenkins Jobs for KubeVirt Lab Validation]({% post_url 2019-10-31-jenkins-jobs-for-kubevirt-lab-validation %}) we covered how Jenkins did get the information about the labs and jobs to perform from the KubeVirt repositories.
+
+In this article, we'll cover the configuration changes in both Jenkins and the JenkinsFiles in order to get our CI setup updated to the latest Jenkins and plugins versions as of this writing.
+
+## Jenkins
+
+Our Jenkins instance, is running on top of [CentOS OpenShift](https://console.apps.ci.centos.org:8443/console/) and is one of the OS-enhanced Jenkins instances, that provide persistent storage and other pieces bundled, required for non-testing setups.
+
+What we found is that Jenkins was already complaining of having lot of updates pending (security, engine, etc), but the `jenkins.war` was embedded in the container image we were using.
```
Comment (ptrnull) [inline _posts/2019-11-22-jenkins-ci-server-upgrade-and-jobs-for-kubevirt.markdown]: ...of having a lot of ...

File: _posts/2019-11-22-jenkins-ci-server-upgrade-and-jobs-for-kubevirt.markdown [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+During the investigation and testing, we found that the persistent storage folder we were using (`/var/lib/jenkins`) contained a `war` folder which contained the unarchived `jenkins.war`, so the next attempt was to manually download the latest `jenkins.war`, and unzip it on that folder, which finally allowed us to upgrade Jenkins core.
+
+## The plugins
+
+After upgrading the Jenkins core, we could use the internal plugin manager to upgrade all the remaining plugins, however, that meant a big change in the plugins, configurations, etc.
+
+After initially being able to run the lab validations for a while, on next morning we wanted to release a new image (from Cloud-Image-Builder) and it failed to build because of the external libraries, and also affected the lab validations again, so we got back to square one for the upgrade process, leaving us with the decision to go forward with the full upgrade: the latest stable jenkins and available plugins and reconfigure what was changed to suit the upgraded requirements.
```
Comment (ptrnull) [inline _posts/2019-11-22-jenkins-ci-server-upgrade-and-jobs-for-kubevirt.markdown]: stable Jenkins and available

File: _posts/2019-11-22-jenkins-ci-server-upgrade-and-jobs-for-kubevirt.markdown [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+### OpenShift Jenkins Sync
+
+Updated and configured to use the console as well with `kubevirt` namespace:
+
+![Jenkins OpenShift sync configuration](/assets/2019-11-22-jenkins-ci-server-upgrade-and-jobs-for-kubevirt/2019-11-11-09-46-47.png)
+
+### Global Pipeline Libraries
+
+Here we added the libraries we used, but instead of specific commit,  targetting the `master` branch.
```
Comment (ptrnull) [inline _posts/2019-11-22-jenkins-ci-server-upgrade-and-jobs-for-kubevirt.markdown]: ... of a specific commit, ...

File: _posts/2019-11-22-jenkins-ci-server-upgrade-and-jobs-for-kubevirt.markdown [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+
+- contra-lib: <https://github.com/openshift/contra-lib.git>
+- cico-pipeline-library: <https://github.com/CentOS/cico-pipeline-library.git>
+- contra-library: <https://github.com/CentOS-PaaS-SIG/contra-env-sample-project>
+
+For all of them, we ticked:
+
+- `Load Implicitly`
+- `Allow default version to be overriden`
```
Comment (ptrnull) [inline _posts/2019-11-22-jenkins-ci-server-upgrade-and-jobs-for-kubevirt.markdown]: ... to be overridden ...

File: _posts/2019-11-22-jenkins-ci-server-upgrade-and-jobs-for-kubevirt.markdown [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+- `Include @Library changes in job recent changes`
+
+Jenkins replied with the 'currently maps to revision: `hash`' for each one of them, after having loaded them properly, indicating that it was successful.
+
+### Slack plugin
+
+In addition to regular plugins used for builds, we incorporated the slack plugin to validate the notifications of build status to a test slack channel. Configuration is really easy, from within Slack, when added the jenkins notifications plugin, a `token` is provided that must be configured in Jenkins as well as a default `room` to send notifications to.
```
Comment (ptrnull) [inline _posts/2019-11-22-jenkins-ci-server-upgrade-and-jobs-for-kubevirt.markdown]: The configuration is really easy...
... when added the Jenkins notifications ...

File: _posts/2019-11-22-jenkins-ci-server-upgrade-and-jobs-for-kubevirt.markdown [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+
+This allows us to get notified when a new build is started and the resulting status, just in case something was generated with errors or something external changed (remember that we use latest kubevirt release, latest tools for virtctl, kubectl and a new image is generated out of them to validate the labs).
```
Comment (ptrnull) [inline _posts/2019-11-22-jenkins-ci-server-upgrade-and-jobs-for-kubevirt.markdown]: ... latest KubeVirt release...

File: _posts/2019-11-22-jenkins-ci-server-upgrade-and-jobs-for-kubevirt.markdown [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+
+![Jenkins-contra-slave container configuration](/assets/2019-11-22-jenkins-ci-server-upgrade-and-jobs-for-kubevirt/2019-11-11-09-54-24.png)
+
+## Other changes
+
+Our environment also used other helper tools as regular OpenShift builds (contra):
+
+![OpenShift Build repositories and status](/assets/2019-11-22-jenkins-ci-server-upgrade-and-jobs-for-kubevirt/2019-11-11-09-55-41.png)
+
+We had to update the repos from using some older forks (no longer valid and outdated) to use the latest versions, and for the ansible-executor we also created a fork to use the newest libraries for accessing Google Cloud environment and tuning some other variables (<https://github.com/CentOS-PaaS-SIG/contra-env-infra/pull/59>) (changes have now landed the upstream repo).
```
Comment (ptrnull) [inline _posts/2019-11-22-jenkins-ci-server-upgrade-and-jobs-for-kubevirt.markdown]: ...the repositories from...

File: _posts/2019-11-22-jenkins-ci-server-upgrade-and-jobs-for-kubevirt.markdown [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+
+The issue that we were facing was related with the failure to write temporary files to the `$HOME` folder for the user so ansible configuration was forced to use one in a temporary and writable folder.
+
+Additionally, Google Cloud access required updating libraries for authentication that were failing as well, that were now fixed in the Dockerfile that generated `ansible-executor` container image.
```
Comment (ptrnull) [inline _posts/2019-11-22-jenkins-ci-server-upgrade-and-jobs-for-kubevirt.markdown]: ...  that was now fixed ...

File: _posts/2019-11-22-jenkins-ci-server-upgrade-and-jobs-for-kubevirt.markdown [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+- Extend the try/catch block to include a 'finally' to send the notifications
+- Change the syntax for 'artifacts' as it was previously `ArchiveArtifacts` and now it's a `postBuild`
+
+## The outcome
+
+After several attempts for fine-tuning the configuration, the builds started succeeding:
+
+![Sunny build status](/assets/2019-11-22-jenkins-ci-server-upgrade-and-jobs-for-kubevirt/2019-11-11-11-02-34.png)
+
+Of course, one of the advantages is that builds happen automatically everyday or on code changes on the repos.
```
Comment (ptrnull) [inline _posts/2019-11-22-jenkins-ci-server-upgrade-and-jobs-for-kubevirt.markdown]: ... every day or on code changes on the repositories.

File: _posts/2019-11-22-jenkins-ci-server-upgrade-and-jobs-for-kubevirt.markdown [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+
+There's still room for improvement identified that will happen in next iterations:
+
+- Find not needed running instances on Cloud providers for reducing the bills
+- Trigger builds when new releases of kubevirt happen (out of kubevirt/kubevirt repo)
```
Comment (ptrnull) [inline _posts/2019-11-22-jenkins-ci-server-upgrade-and-jobs-for-kubevirt.markdown]: ... of KubeVirt ...

File: _posts/2019-11-22-jenkins-ci-server-upgrade-and-jobs-for-kubevirt.markdown [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+title: Jenkins Infra upgrade
+navbar_active: Blogs
+pub-date: Nov 22
+pub-year: 2019
+category: news
+---
+
+## Introduction
+
+In the article [Jenkins Jobs for KubeVirt Lab Validation]({% post_url 2019-10-31-jenkins-jobs-for-kubevirt-lab-validation %}) we covered how Jenkins did get the information about the labs and jobs to perform from the KubeVirt repositories.
```
[the remark this answers is not in the document]
Comment (iranzo) [reply _posts/2019-11-22-jenkins-ci-server-upgrade-and-jobs-for-kubevirt.markdown]: done in next push

**Pull request merged**